### PR TITLE
Add the homogenised Laplace operator

### PR DESCRIPTION
**Pull request by removr**
Implements the homogenised Laplace operator with the help of solid spherical harmonics. The coefficients of the expansion are calculated with Eigen.

### CODE REVIEW

File: Bembel/src/HomogenisedLaplace/SingleLayerPotential.hpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * \ingroup HomogenisedLaplace
+ */
+template<typename LinOp>
+class HomogenisedLaplaceSingleLayerPotential: public PotentialBase<
+    HomogenisedLaplaceSingleLayerPotential<LinOp>, LinOp> {
+  // implementation of the kernel evaluation, which may be based on the
+  // information available from the superSpace
+
+ private:
```
Comment (mx-nlte) [inline Bembel/src/HomogenisedLaplace/SingleLayerPotential.hpp:36]: (optional) private members are at the bottom of the class definition in Bembel.

File: Bembel/src/HomogenisedLaplace/SingleLayerPotential.hpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        HomogenisedLaplaceSingleLayerOperator::getPrecision());
+  }
+  Eigen::Matrix<
+      typename PotentialReturnScalar<
+          typename LinearOperatorTraits<LinOp>::Scalar,
+            double>::Scalar, 1, 1> evaluateIntegrand_impl(
+      const FunctionEvaluator<LinOp> &fun_ev, const ElementTreeNode &element,
+      const Eigen::Vector3d &point, const SurfacePoint &p) const {
+    // get evaluation points on unit square
+    auto s = p.segment < 2 > (0);
```
Comment (mx-nlte) [inline Bembel/src/HomogenisedLaplace/SingleLayerPotential.hpp:60]: Please remove the auto.

File: examples/HomogenisedLaplaceSingleLayerFull.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+#include <Bembel/AnsatzSpace>
+#include <Bembel/Geometry>
+#include <Bembel/H2Matrix>
+#include <Bembel/HomogenisedLaplace>
+#include <Bembel/LinearForm>
+
+#include <iostream>
+
+#include "./Data.hpp"
```
Comment (mx-nlte) [inline examples/HomogenisedLaplaceSingleLayerFull.cpp:22]: Replace '.' with 'examples'

File: examples/HomogenisedLaplaceSingleLayerFull.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      // evaluate potential
+      Bembel::DiscretePotential<
+          HomogenisedLaplaceSingleLayerPotential<
+              HomogenisedLaplaceSingleLayerOperator>,
+          HomogenisedLaplaceSingleLayerOperator> disc_pot(ansatz_space);
+      disc_pot.set_cauchy_data(rho);
+      VectorXd pot = disc_pot.evaluate(gridpoints);
+
+      // print error
+      std::cout << Bembel::maxPointwiseError<double>(pot, gridpoints, fun)
```
Comment (mx-nlte) [inline examples/HomogenisedLaplaceSingleLayerFull.cpp:92]: Do we expect a certain convergence order? We have macros to check that in 'examples/Error.hpp'.

File: geo/cube_small.dat [hunk 5]
```diff
@@ -0,0 +1,107 @@
+# nurbs mesh v.2.1
+#
+# 09-Jan-2023
```
Comment (mx-nlte) [inline geo/cube_small.dat:3]: (optional) remove the date

File: geo/cube_small.dat [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+PATCH 6 
+1 1 
+2 2 
+0.000000000000000   0.000000000000000   1.000000000000000   1.000000000000000   
+0.000000000000000   0.000000000000000   1.000000000000000   1.000000000000000   
+0.000000000000000   0.250000000000000   0.000000000000000   0.250000000000000   
+0.000000000000000   0.000000000000000   0.250000000000000   0.250000000000000   
+0.250000000000000   0.250000000000000   0.250000000000000   0.250000000000000   
+1.000000000000000   1.000000000000000   1.000000000000000   1.000000000000000   
+INTERFACE 1
```
Comment (mx-nlte) [inline geo/cube_small.dat:60]: (optional) we do not need the interfaces. For consistency we should remove them.

File: tests/test_HomogenisedCoefficients.cpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+// provided WITHOUT ANY WARRANTY, see <http://www.bembel.eu> for further
+// information.
+
+#include <Bembel/HomogenisedLaplace>
+#include <Eigen/Dense>
+
+#include <iostream>
+#include <functional>
+
+#include "./Test.hpp"
```
Comment (mx-nlte) [inline tests/test_HomogenisedCoefficients.cpp:18]: replace '.' with 'tests'

File: Bembel/HomogenisedLaplace [hunk 8]
```diff
@@ -0,0 +1,25 @@
+// This file is part of Bembel, the higher order C++ boundary element library.
+// It was written as part of a cooperation of J. Doelz, H. Harbrecht, S. Kurz,
```
Comment (mx-nlte) [inline Bembel/HomogenisedLaplace:2]: Replace with new header including the copy right.

File: Bembel/src/HomogenisedLaplace/Coefficients.hpp [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+
+#ifndef BEMBEL_SRC_HOMOGENISEDLAPLACE_COEFFICIENTS_HPP_
+#define BEMBEL_SRC_HOMOGENISEDLAPLACE_COEFFICIENTS_HPP_
+
+#ifndef POINT_DEGREE
+#define POINT_DEGREE 20 /** the number of points on the surface of the cube */
+#endif
+
+#ifndef PI
+#define PI M_PI
```
Comment (mx-nlte) [inline Bembel/src/HomogenisedLaplace/Coefficients.hpp:20]: Please use M_PI directly.

File: Bembel/src/HomogenisedLaplace/SingleLayerOperator.hpp [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+};
+
+/**
+ * \ingroup HomogenisedLaplace
+ */
+class HomogenisedLaplaceSingleLayerOperator: public LinearOperatorBase<
+    HomogenisedLaplaceSingleLayerOperator> {
+  // implementation of the kernel evaluation, which may be based on the
+  // information available from the superSpace
+ private:
```
Comment (mx-nlte) [inline Bembel/src/HomogenisedLaplace/SingleLayerOperator.hpp:38]: Here are the same points as in the SingleLayerPotential.hpp

File: examples/HomogenisedLaplaceSingleLayerH2.cpp [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+#include <Bembel/AnsatzSpace>
+#include <Bembel/Geometry>
+#include <Bembel/H2Matrix>
+#include <Bembel/HomogenisedLaplace>
+#include <Bembel/LinearForm>
+
+#include <iostream>
+
+#include "./Data.hpp"
```
Comment (mx-nlte) [inline examples/HomogenisedLaplaceSingleLayerH2.cpp:22]: Same as in the other example

**Review by mx-nlte — COMMENTED** [review]
I like your Doxygen comments

File: Bembel/HomogenisedLaplace (hunk 8 above)
Comment (removr) [reply Bembel/HomogenisedLaplace:5]: Done

File: Bembel/src/HomogenisedLaplace/Coefficients.hpp (hunk 9 above)
Comment (removr) [reply Bembel/src/HomogenisedLaplace/Coefficients.hpp]: Done

File: Bembel/src/HomogenisedLaplace/SingleLayerPotential.hpp (hunk 1 above)
Comment (removr) [reply Bembel/src/HomogenisedLaplace/SingleLayerPotential.hpp]: Done

File: Bembel/src/util/Sphericals.hpp [hunk 12]
```diff
@@ -0,0 +1,767 @@
...
+// source code is subject to the GNU General Public License version 3 and
+// provided WITHOUT ANY WARRANTY, see <http://www.bembel.eu> for further
+// information.
+
+#ifndef BEMBEL_SRC_UTIL_SPHERICALS_HPP_
+#define BEMBEL_SRC_UTIL_SPHERICALS_HPP_
+#include <Eigen/Dense>
+
+#if !defined pi
+#define pi 3.1415926535897932385
```
[the remark this answers is not in the document]
Comment (removr) [reply Bembel/src/util/Sphericals.hpp]: Done

File: Bembel/src/util/Sphericals.hpp [hunk 13]
```diff
@@ -0,0 +1,767 @@
...
+ * 																			  (2, -2), (2, -1), ... (n, n)] \f$
+ * @param deg:	The degree
+ */
+double evaluate_sphericals(Eigen::Vector3d x, Eigen::VectorXd cs,
+    unsigned int deg) {
+  unsigned int m, n;
+  double z1[2], z2[2], z3[2], z1_start[2];
+  double r, fac, rootTimesZ, root_2, root_3;
+
+  assert(abs(x.norm() - 1) < 1e-14);
```
[the remark this answers is not in the document]
Comment (removr) [reply Bembel/src/util/Sphericals.hpp]: Done

File: Bembel/src/util/Sphericals.hpp [hunk 14]
```diff
@@ -0,0 +1,767 @@
...
+ * @param deg:	The degree
+ */
+Eigen::Vector3d evaluate_dsphericals(Eigen::Vector3d x, Eigen::VectorXd cs,
+    unsigned int deg) {
+  unsigned int m, n;
+  double z1[2], z2[2], z3[2], z1_start[2];
+  Eigen::Vector3d dr;
+  double fac, rootTimesZ, root_2, root_3;
+
+  assert(abs(x.norm() - 1) < 1e-14);
```
[the remark this answers is not in the document]
Comment (removr) [reply Bembel/src/util/Sphericals.hpp]: Done

File: examples/HomogenisedLaplaceSingleLayerFull.cpp (hunk 3 above)
Comment (removr) [reply examples/HomogenisedLaplaceSingleLayerFull.cpp]: Done

File: tests/test_HomogenisedCoefficients.cpp (hunk 7 above)
Comment (removr) [reply tests/test_HomogenisedCoefficients.cpp]: Done

File: examples/HomogenisedLaplaceSingleLayerFull.cpp (hunk 4 above)
Comment (removr) [reply examples/HomogenisedLaplaceSingleLayerFull.cpp]: As far as I know, the same order as for the "common" single layer operator. This is now included in the example file.

File: examples/HomogenisedLaplaceSingleLayerH2.cpp (hunk 11 above)
Comment (removr) [reply examples/HomogenisedLaplaceSingleLayerH2.cpp]: Done

**Review by mx-nlte — APPROVED** [review]

**Pull request merged**